src/control/lib/control/system.go

```diff
@@ -50,6 +52,36 @@ type sysResponse struct {
 	AbsentHosts hostlist.HostSet
 }
 
+func (sr *sysResponse) getAbsentHostsRanks(inHosts, inRanks string) error {
+	ahs, err := hostlist.CreateSet(inHosts)
+	if err != nil {
+		return err
+	}
+	ars, err := system.CreateRankSet(inRanks)
+	if err != nil {
+		return err
+	}
+	sr.AbsentHosts = *ahs
+	sr.AbsentRanks = *ars
+
+	return nil
+}
+
+func (sr *sysResponse) DisplayAbsentHostsRanks() string {
+	switch {
+	case sr.AbsentHosts.Count() > 0:
+		return fmt.Sprintf("\nUnknown %s: %s",
+			english.Plural(sr.AbsentHosts.Count(), "host", "hosts"),
+			sr.AbsentHosts.String())
+	case sr.AbsentRanks.Count() > 0:
+		return fmt.Sprintf("\nUnknown %s: %s",
+			english.Plural(sr.AbsentRanks.Count(), "rank", "ranks"),
+			sr.AbsentRanks.String())
+	default:
+		return ""
+	}
+}
```

Comment:
I'm not crazy about adding this display logic to the response struct. I wouldn't block on it but I'm concerned about setting the precedent. The response structs should ideally be agnostic about display, leaving those decisions to API callers (which may just dump the responses as JSON).

In fact, I'm curious as to whether or not you've tested all of this with the JSON output option. We need to make sure that format works OK (i.e. provides the same data but in machine-parsable format).

Reply:
I did have this method in `dmg/system.go` but moved it to the response so I didn't have to pass the absent values in, I will move it back out again in a subsequent PR as I agree with the premise that display logic shouldn't be part of the control API.

I don't think the changes have interfered with JSON output:
```
bash-4.2$ dmg system query --rank-hosts wolf-7[1-9] --json
{
  "AbsentRanks": "",
  "AbsentHosts": "wolf-[73-79]",
  "Members": [
    {
      "Addr": "10.8.1.71:10001",
      "State": 4,
      "Rank": 0,
      "UUID": "7d1f163f-f950-486b-bcfa-b05a200abaf3",
      "Info": ""
    },
    {
      "Addr": "10.8.1.72:10001",
      "State": 9,
      "Rank": 1,
      "UUID": "30cf419a-9a1c-44b1-b568-e7d74b9566c8",
      "Info": "client: code = 533 description = \"the server at 10.8.1.72:10001 refused the connection\""
    },
    {
      "Addr": "10.8.1.72:10001",
      "State": 9,
      "Rank": 2,
      "UUID": "8d1fa14a-22e0-4c05-9650-56c80d6118ff",
      "Info": "client: code = 533 description = \"the server at 10.8.1.72:10001 refused the connection\""
    },
    {
      "Addr": "10.8.1.71:10001",
      "State": 4,
      "Rank": 3,
      "UUID": "f4e3b01f-d2c2-4f89-83f6-48b2ac19c55d",
      "Info": ""
    }
  ]
}
bash-4.2$ dmg system start --rank-hosts wolf-7[1-9] --json
 {
  "AbsentRanks": "",
  "AbsentHosts": "wolf-[73-79]",
  "Results": [
    {
      "State": 3,
      "Addr": "10.8.1.71:10001",
      "Rank": 0,
      "Action": "start",
      "Errored": false,
      "Msg": "system start"
    },
    {
      "State": 3,
      "Addr": "10.8.1.71:10001",
      "Rank": 3,
      "Action": "start",
      "Errored": false,
      "Msg": "system start"
    },
    {
      "State": 9,
      "Addr": "10.8.1.72:10001",
      "Rank": 1,
      "Action": "start",
      "Errored": false,
      "Msg": "client: code = 533 description = \"the server at 10.8.1.72:10001 refused the connection\""
    },
    {
      "State": 9,
      "Addr": "10.8.1.72:10001",
      "Rank": 2,
      "Action": "start",
      "Errored": false,
      "Msg": "client: code = 533 description = \"the server at 10.8.1.72:10001 refused the connection\""
    }
  ]
}
bash-4.2$
```

Reply:
In your follow-on PR, I would like to see these response structures reworked a bit to bring them in line with the other response types. In particular, errors should be in a `HostErrorsMap`, like every other response type. This will keep the API consistent.

As it is currently, API callers have to pick through all of the results to find the errors, which is not ideal.

Reply:
the membership currently works by receiving an aggregated list of results which get operated on in order to update the membership appropriately. that's why the system commands return results for each rankk (member results).  If you want me to revise that design . please put some detail in a ticket and I will take a look. these don't return HostErrorsMap because they are msRequest responses.